File: miscellaneous/sorting/merge-sort/20230103/merge-sort.js
'use strict';

/**
 *
 * @param nums [number]
 * @returns {*}
 */
function mergeSort(nums) {
    if (nums.length <= 1) return nums;

    const mid = Math.floor((nums.length) / 2);

    return merge(
        mergeSort(nums.slice(0, mid)),
        mergeSort(nums.slice(mid))
    );
}

function merge(left, right) {
    const merged = [];

    while (left.length && right.length) {
        merged.push(left[0] > right[0] ? right.shift() : left.shift());
    }

    return [...merged, ...left, ...right];
}

module.exports = mergeSort;

File: miscellaneous/sorting/merge-sort/20230103/merge_sort.py
from typing import List
from math import floor

def merge_sort(nums: List[int]) -> List[int]:
    if len(nums) <= 1:
        return nums

    mid = floor(len(nums)/2)

    merged = merge(merge_sort(nums[:mid]), merge_sort(nums[mid:]))
    return merged


def merge(left: List[int], right: List[int]) -> List[int]:
    merged: List[int] = []

    while len(left) and len(right):
        merged.append(left.pop(0) if left[0] < right[0] else right.pop(0))

    return merged + left + right